components/raftstore-v2/src/operation/command/admin/commit_merge.rs

```diff
@@ -0,0 +1,927 @@
+            return Err(box_err!(
+                "peers not matched: {:?} {:?}",
+                source_region,
+                region
+            ));
+        }
+        let mut proposal_ctx = ProposalContext::empty();
+        proposal_ctx.insert(ProposalContext::COMMIT_MERGE);
+        let data = req.write_to_bytes().unwrap();
+        self.propose_with_ctx(store_ctx, data, proposal_ctx.to_vec())
+    }
+}
+
+impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
+    // Match v1::exec_commit_merge.
+    pub async fn apply_commit_merge(
+        &mut self,
+        req: &AdminRequest,
+        index: u64,
+    ) -> Result<(AdminResponse, AdminCmdResult)> {
+        PEER_ADMIN_CMD_COUNTER.commit_merge.all.inc();
+
+        self.flush();
+
+        // Note: doesn't validate region state from kvdb any more.
+        let merge = req.get_commit_merge();
+        let source_region = merge.get_source();
+        let (tx, rx) = oneshot::channel();
+        self.res_reporter().send(PeerMsg::CatchUpLogs(CatchUpLogs {
+            target_region_id: self.region_id(),
+            merge: merge.clone(),
+            tx,
+        }));
+        let source_ready = rx.await.unwrap_or_else(|_| {
+            // TODO: handle this gracefully.
+            slog_panic!(self.logger, "source peer is missing");
+        });
+
+        info!(
+            self.logger,
+            "execute CommitMerge";
+            "commit" => merge.get_commit(),
+            "entries" => merge.get_entries().len(),
+            "index" => index,
+            "source_region" => ?source_region
+        );
+
+        let mut region = self.region().clone();
+        // Use a max value so that pd can ensure overlapped region has a priority.
+        let version = cmp::max(
+            source_region.get_region_epoch().get_version(),
+            region.get_region_epoch().get_version(),
+        ) + 1;
+        region.mut_region_epoch().set_version(version);
+        if keys::enc_end_key(&region) == keys::enc_start_key(source_region) {
+            region.set_end_key(source_region.get_end_key().to_vec());
+        } else {
+            region.set_start_key(source_region.get_start_key().to_vec());
+        }
+
+        let source_tablet: EK = match source_ready.tablet.downcast() {
+            Ok(t) => *t,
+            Err(t) => unreachable!("tablet type should be the same: {:?}", t),
+        };
+        self.tablet().flush_cfs(&[], true).unwrap();
+        // TODO: check both are trimmed.
+        let reg = self.tablet_registry();
+        let tmp_path = reg.tablet_path(self.region_id(), index);
+        if tmp_path.exists() {
+            std::fs::remove_dir_all(&tmp_path).unwrap();
+        }
+        let mut ctx = TabletContext::new(&region, Some(index));
+        {
+            let tablet = reg
+                .tablet_factory()
+                .open_tablet(ctx.clone(), &tmp_path)
+                .unwrap();
+            tablet.merge(&[&source_tablet, self.tablet()]).unwrap();
```

Comment:
Can merge be done in checkpoint? It may be a lot safer as only manifest needs to be changed.

Reply:
Checkpoint has limited functionality. It only copies file without altering the content (except for the CURRENT which is a plaintext marker file). It also cannot resolve conflict of file number between instances.

Reply:
What I mean is create a checkpoint and then modify it.